shark_turbine/kernel/wave/expansion.py

```diff
@@ -208,14 +214,40 @@ def expand_graph(
                 dim: val for dim, val in zip(dim_scaling.keys(), dim_combination)
             }
             logger.debug(f"Starting expansion at leaf:{node} in dims:{expand_dims}")
-            _expand_node(node, trace, expand_dims, dim_scaling, expansion_context)
+            _expand_node(
+                node, trace, expand_dims, dim_scaling, dim_tile_size, expansion_context
+            )
+
+    # Replace symbolic tile sizes with actual tile sizes.
+    def replace_index(node: fx.Node) -> bool:
+        if node.index is not None:
+            for dim in node.index:
+                node.index[dim].start = node.index[dim].start.subs(tile_size_mapping)
+        return False
+
+    trace.walk(replace_index)
```

Comment:
Why is this part of the expansion? Is there no case where we might want these symbolic later?
Does this have to be done by walking all nodes or could this be done when we visit the nodes individually?

Reply:
I had the same question since I was expecting this to be part of the `codegen.py`, but here is Harsh's response:

>  But the reason why we dont want to add this into indexing_context is because this is temporary index only required during the expansion phase.

Reply:
So this brings up a good question. My naive attempt at indexing was to just update the index but when I did this I saw overcounting happening which implied that nodes are being visited multiple times. Do you know why this happens? So my understanding is that if M = [0, 1, 2] and N = [0, 1], then we should visit the node only 3 x 2 = 6 times, but we are visiting it more than that. Why is this the case? 

The reason why this is part of the expansion is because I dont think it makes sense to keep the tile sizes symbolic after expansion is materialized. Not sure what the benefit of that would be. I dont think we can do this by visiting the nodes individually unless we figure out the first question - why do we multiple a node more than its dim scaling requires it to?

Reply:
> My naive attempt at indexing was to just update the index but when I did this I saw overcounting happening which implied that nodes are being visited multiple times

Oof, that looks like we have a bug in the check for whether or not we have expanded the node in that dimension? 


IIRC we have a piece of code that queries the map who have `original_node+Indexing -> expanded_node` to check that we have expanded before. as seen on https://github.com/iree-org/iree-turbine/blob/f0271e8e55fb1d6e51236a621b17a1ff97c1ccf5/shark_turbine/kernel/wave/expansion.py#L221-L225

Reply:
this is no longer required. I have reworked the code.